ext/cl_khr_command_buffer_mutable_dispatch.asciidoc

```diff
@@ -533,9 +561,9 @@ the array violates the defined conditions:
 * {CL_INVALID_VALUE} if _type_ is not
   {CL_STRUCTURE_TYPE_MUTABLE_DISPATCH_CONFIG_KHR}.
 
-* {CL_INVALID_OPERATION} if values of _local_work_size_ and/or
-  _global_work_size_ result in an increase to the number of work-groups in the
-  ND-range.
+* {CL_INVALID_OPERATION} if {CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR} is specified, 
+  and values of _local_work_size_ and/or _global_work_size_ result in an increase to the number of work-
+  groups in the ND-range over the number specified when the ND-range kernel was recorded.
```

Comment:
What is the expected behavior if a command buffer or command specifies `CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR`, the local work size is initially non-NULL, and then the updated local work size **is** NULL?

Reply:
Hmm, I haven't thought of that. I think that, in this case, we should allow this, and the implementation should be able to handle it.

Reply:
Hmm, perhaps there are two scenarios to consider:

1. The global work size is not changing, but the local work size is changing to NULL.  I this case I suppose that the implementation could handle it, although it may need to choose a different local work size than it would otherwise, subject to the "no additional work-groups" constraint.

2. The global work size is changing, and the local work size is also changing to NULL.  I think it would be very difficult to handle this in a general case, though, and there may be some global work sizes that satisfy the "no additional work-groups" constraint.
    * Note that this is true even if the global work size gets smaller, especially if the work-group size must be uniform; for example, consider changing from a very large power-of-two (with lots of prime factors and options for work-group sizes) to a slightly smaller prime number (with few work-group size options).

I see that the latest updates added the line:

```
If the user updates _local_work_size_ to be _NULL_, the assertion is ignored.
```

We may want to go a bit stronger though, and either make this an error (assuming we're checking errors), or undefined behavior (if we aren't defining errors).

Reply:
Sorry about the false alarm, this may not actually be an issue.

In the current mutable dispatch extension we say (**emphasis** mine):

> local_work_size Points to an array of work_dim unsigned values that describe the number of work-items that make up a work-group that will execute the kernel. **If local_work_size is NULL then the number of local work-items in the dispatch is not changed.** See clEnqueueNDRangeKernel for valid usage.

This means that it is not possible to change from a non-NULL local work size to a NULL local work size, because the NULL local work size will be interpreted as a request to keep the local work size unchanged.  Therefore, the case where the "user updates _local_work_size_ to be _NULL_" will never happen, and we don't need to include this case in the spec, although I suppose we could include it in a footnote or an informative note, if we so choose.

The situations we need to consider are for the assertion are:

1. Non-NULL initial local work size, non-NULL updated local work size.  In this case the updated work-group count is computed from the updated local work size and must be less than or equal to the initial work-group count or behavior is undefined.
2. Non-NULL initial local work size and "NULL" updated local work size.  In this case, the work-group count is computed from the initial local work size (since the "NULL" local work size indicates that it is unchanged), and the updated work-group count must be less than or equal to the initial work-group count or behavior is undefined.
3. NULL initial local work size.  For this case, the PR currently says "In case _local_work_size_ is NULL, the effect of this flag is undefined", so I suppose behavior is undefined in all of these scenarios.  Can we do any better?  Some possibilities:
    * The assertion is ignored when the local work size is NULL.
    * The assertion is invalid when the local work size is NULL.  Note, this condition would be checked when the command buffer is initially recorded, not whenever it is modified, so error checking should be possible.  (I prefer this option.)
    * The initial work-group count is computed assuming a worst-case work-group size of one.  (This seems a little too clever.).

Does this sound correct?

Reply:
You are right; I also forgot about this part of the spec.
I, too, think that the 2nd option is preferable. I'll update the PR accordingly.